processor/stream/processor.go

```diff
@@ -272,6 +277,23 @@ func (p *Processor) getStreamReader(r io.Reader) *streamReader {
 	}
 }
 
+func (p *Processor) acquireLock(ctx context.Context, d time.Duration) (func(), error) {
+	select {
+	case p.sem <- struct{}{}:
+	default:
+		t := acquireTimer(d)
+		defer releaseTimer(t)
```

Comment:
Did you run some benchmarks that found the pooling to be worthwhile? I wonder if we should keep it simple and use `context.WithTimeout`, and then check for `context.DeadlineExceeded` and convert that to `publish.ErrFull`?

Reply:
that may be better indeed. I didn't benchmark the change and I think it may be too much to begin with, but since It's in the hot path, I wanted to reuse the timers, since `time.After` will add to the number of GC objects, but that may not be a problem since the `time.Timer` looks like a pretty small `struct`.

I can run some benchmarks and see how it compares.

If it turns out the overhead is pretty small and pooling isn't worth it, do you have a preference between using `context.WithTimeout` and `time.After`? The latter seems much more contained than the former, so it may be slightly less wasteful to use `time.After`?

Reply:
I suppose I would slightly prefer a timer, as the context could in theory already have a deadline which should not be converted to ErrFull.

Reply:
did some benchmarks and unsurprisingly the difference isn't massive, but at first the difference between the pooled and the `time.After` approach was bigger than I expected:

```
$ benchstat pool.txt timeafter.txt
name                             old time/op    new time/op    delta
BackendProcessorParallelBlock-8    6.62ms ± 2%    6.84ms ± 1%  +3.25%  (p=0.000 n=10+9)

name                             old speed      new speed      delta
BackendProcessorParallelBlock-8  60.2MB/s ± 2%  58.3MB/s ± 1%  -3.15%  (p=0.000 n=10+9)

name                             old alloc/op   new alloc/op   delta
BackendProcessorParallelBlock-8    13.5MB ± 0%    13.6MB ± 0%  +0.01%  (p=0.010 n=10+9)

name                             old allocs/op  new allocs/op  delta
BackendProcessorParallelBlock-8     24.8k ± 0%     24.8k ± 0%  +0.03%  (p=0.000 n=10+10)
```

This is the benchmark I used:

```go
func BenchmarkBackendProcessorParallelBlock(b *testing.B) {
	cfg := config.DefaultConfig()
	cfg.MaxConcurrentDecoders = 1
	processor := BackendProcessor(cfg, make(chan struct{}, cfg.MaxConcurrentDecoders))
	file := "../../testdata/intake-v2/heavy.ndjson"
	const batchSize = 10
	batchProcessor := nopBatchProcessor{} // This is probably unrealistic since the processors will take time to process each event.
	data, err := ioutil.ReadFile(file)
	if err != nil {
		b.Error(err)
	}
	b.SetBytes(int64(len(data)))

	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()
	b.ResetTimer()
	b.RunParallel(func(p *testing.PB) {
		r := bytes.NewReader(data)
		for p.Next() {
			var result Result
			processor.HandleStream(ctx, model.APMEvent{}, r, batchSize, batchProcessor, &result)
			r.Seek(0, io.SeekStart)
		}
	})
}
```

Then I decided to write a small mock processor that has 2 muteness similar to what the model indexer has:

```go
type mutexBatchProcessor struct {
	rwm sync.RWMutex
	mu  sync.Mutex
}

func (p *mutexBatchProcessor) ProcessBatch(_ context.Context, b *model.Batch) error {
	p.rwm.RLock()
	defer p.rwm.RUnlock()
	// Simulate modelindexer
	for range *b {
		p.mu.Lock()
		p.mu.Unlock()
	}
	return nil
}
```

Using that `mutexBatchProcessor` in the above benchmark resulted in a smaller difference in performance, but the pooled approach still looks like it's generally faster:

```
$ benchstat pool-mi-sim.txt timeafter-mi-sim.txt
name                             old time/op    new time/op    delta
BackendProcessorParallelBlock-8    6.62ms ± 2%    6.80ms ± 1%  +2.63%  (p=0.000 n=10+9)

name                             old speed      new speed      delta
BackendProcessorParallelBlock-8  60.2MB/s ± 2%  58.7MB/s ± 1%  -2.57%  (p=0.000 n=10+9)

name                             old alloc/op   new alloc/op   delta
BackendProcessorParallelBlock-8    13.5MB ± 0%    13.5MB ± 0%    ~     (p=0.739 n=10+10)

name                             old allocs/op  new allocs/op  delta
BackendProcessorParallelBlock-8     24.8k ± 0%     24.8k ± 0%    ~     (p=0.190 n=10+10)
```

Then I tried to increase the number of GC'd and re-used `time.Timer` by setting `-cpu=50`, the deltas were pretty similar, though:

```
benchstat pool-mi-sim-50-cpu.txt timeafter-mi-sim-50-cpu.txt
name                              old time/op    new time/op    delta
BackendProcessorParallelBlock-50    7.00ms ± 2%    7.16ms ± 1%  +2.34%  (p=0.000 n=9+10)

name                              old speed      new speed      delta
BackendProcessorParallelBlock-50  57.0MB/s ± 2%  55.7MB/s ± 1%  -2.28%  (p=0.000 n=9+10)

name                              old alloc/op   new alloc/op   delta
BackendProcessorParallelBlock-50    13.7MB ± 0%    13.7MB ± 0%    ~     (p=0.579 n=10+10)

name                              old allocs/op  new allocs/op  delta
BackendProcessorParallelBlock-50     24.8k ± 0%     24.8k ± 0%  +0.01%  (p=0.000 n=9+10)
```

----

I then increased the capacity of the semaphore to `10` and set the `-cpu=100` to increase concurrency and see how it fares then, the differences were much smaller, so small that I'm not sure if they'll have any effect in production:

```
name                               old time/op    new time/op    delta
BackendProcessorParallelBlock-100    2.74ms ± 1%    2.76ms ± 2%  +0.71%  (p=0.046 n=9+8)

name                               old speed      new speed      delta
BackendProcessorParallelBlock-100   146MB/s ± 1%   145MB/s ± 2%  -0.70%  (p=0.043 n=9+8)

name                               old alloc/op   new alloc/op   delta
BackendProcessorParallelBlock-100    13.7MB ± 0%    13.7MB ± 0%  -0.10%  (p=0.000 n=10+8)

name                               old allocs/op  new allocs/op  delta
BackendProcessorParallelBlock-100     24.8k ± 0%     24.8k ± 0%    ~     (p=0.983 n=10+8)
```

After seeing this maybe it's worth just using `time.After` and fix it in the future if it's broken.

Reply:
Thanks, looks good. We could add small optimisation to use `time.NewTimer` & `defer timer.Stop()` to GC the timer before the timer fires (i.e. if the queue opens back up). Not a big deal though.